Extraneous comma in trait bound suggestion

The following code ([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=87ae1439a80eba5fe72b38e12301d29c)):

```rust
trait T1 {
    type A;
}

trait T2: T1
where
    Self::A: Clone,
{
}

fn generic<T: T2>()
where
    u8: Copy,
{
}
```

Produces an invalid suggestion:

```
error[E0277]: the trait bound `<T as T1>::A: std::clone::Clone` is not satisfied
  --> src/lib.rs:11:15
   |
5  | trait T2: T1
   |       -- required by a bound in this
6  | where
7  |     Self::A: Clone,
   |              ----- required by this bound in `T2`
...
11 | fn generic<T: T2>()
   |               ^^ the trait `std::clone::Clone` is not implemented for `<T as T1>::A`
   |
help: consider further restricting the associated type
   |
13 |     u8: Copy,, <T as T1>::A: std::clone::Clone
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

on nightly. On stable, the error message looks a bit different but seems to also be suggesting an extra comma.

When trying to reduce this test case I found that in other cases this same lint seems to suggest the right thing (this may be useful as an additional test case):

```rust
#[derive(Clone)]
struct Generic<T>(T);

fn generic<T>() -> impl Clone
where
    Generic<T>: Default,
{
    Generic::<T>::default()
}
```

correct suggestion (nightly):

```
Standard Error

   Compiling playground v0.0.1 (/playground)

error[E0277]: the trait bound `T: std::clone::Clone` is not satisfied
 --> src/lib.rs:4:20
  |
4 | fn generic<T>() -> impl Clone
  |                    ^^^^^^^^^^ the trait `std::clone::Clone` is not implemented for `T`
...
8 |     Generic::<T>::default()
  |     ----------------------- this returned value is of type `Generic<T>`
  |
  = note: required because of the requirements on the impl of `std::clone::Clone` for `Generic<T>`
  = note: the return type of a function must have a statically known size
help: consider further restricting type parameter `T`
  |
6 |     Generic<T>: Default, T: std::clone::Clone
  |                        ^^^^^^^^^^^^^^^^^^^^^^
```

--- a/src/librustc_hir/hir.rs
+++ b/src/librustc_hir/hir.rs
@@ -525,6 +525,13 @@ impl WhereClause<'_> {
     pub fn span_for_predicates_or_empty_place(&self) -> Span {
         self.span
     }
+
+    /// `Span` where further predicates would be suggested, accounting for trailing commas, like
+    ///  in `fn foo<T>(t: T) where T: Foo,` so we don't suggest two trailing commas.
+    pub fn tail_span_for_suggestion(&self) -> Span {
+        let end = self.span_for_predicates_or_empty_place().shrink_to_hi();
+        self.predicates.last().map(|p| p.span()).unwrap_or(end).shrink_to_hi().to(end)
+    }
 }
 
 /// A single predicate in a where-clause.

--- a/src/librustc_middle/ty/diagnostics.rs
+++ b/src/librustc_middle/ty/diagnostics.rs
@@ -220,22 +220,11 @@ pub fn suggest_constraining_type_param(
             }
         }
 
-        // Account for `fn foo<T>(t: T) where T: Foo,` so we don't suggest two trailing commas.
-        let end = generics.where_clause.span_for_predicates_or_empty_place().shrink_to_hi();
-        let where_clause_span = generics
-            .where_clause
-            .predicates
-            .last()
-            .map(|p| p.span())
-            .unwrap_or(end)
-            .shrink_to_hi()
-            .to(end);
-
         match &param_spans[..] {
             &[&param_span] => suggest_restrict(param_span.shrink_to_hi()),
             _ => {
                 err.span_suggestion_verbose(
-                    where_clause_span,
+                    generics.where_clause.tail_span_for_suggestion(),
                     &msg_restrict_type_further,
                     format!(", {}: {}", param_name, constraint),
                     Applicability::MachineApplicable,

--- a/src/librustc_middle/ty/diagnostics.rs
+++ b/src/librustc_middle/ty/diagnostics.rs
@@ -7,7 +7,6 @@ use rustc_errors::{Applicability, DiagnosticBuilder};
 use rustc_hir as hir;
 use rustc_hir::def_id::DefId;
 use rustc_hir::{QPath, TyKind, WhereBoundPredicate, WherePredicate};
-use rustc_span::{BytePos, Span};
 
 impl<'tcx> TyS<'tcx> {
     /// Similar to `TyS::is_primitive`, but also considers inferred numeric values to be primitive.
@@ -221,18 +220,16 @@ pub fn suggest_constraining_type_param(
             }
         }
 
-        let where_clause_span = generics.where_clause.span_for_predicates_or_empty_place();
         // Account for `fn foo<T>(t: T) where T: Foo,` so we don't suggest two trailing commas.
-        let mut trailing_comma = false;
-        if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(where_clause_span) {
-            trailing_comma = snippet.ends_with(',');
-        }
-        let where_clause_span = if trailing_comma {
-            let hi = where_clause_span.hi();
-            Span::new(hi - BytePos(1), hi, where_clause_span.ctxt())
-        } else {
-            where_clause_span.shrink_to_hi()
-        };
+        let end = generics.where_clause.span_for_predicates_or_empty_place().shrink_to_hi();
+        let where_clause_span = generics
+            .where_clause
+            .predicates
+            .last()
+            .map(|p| p.span())
+            .unwrap_or(end)
+            .shrink_to_hi()
+            .to(end);
 
         match &param_spans[..] {
             &[&param_span] => suggest_restrict(param_span.shrink_to_hi()),

--- a/src/librustc_trait_selection/traits/error_reporting/suggestions.rs
+++ b/src/librustc_trait_selection/traits/error_reporting/suggestions.rs
@@ -169,17 +169,8 @@ pub trait InferCtxtExt<'tcx> {
 }
 
 fn predicate_constraint(generics: &hir::Generics<'_>, pred: String) -> (Span, String) {
-    let end = generics.where_clause.span_for_predicates_or_empty_place().shrink_to_hi();
     (
-        // Account for `where T: Foo,` so we don't suggest two trailing commas.
-        generics
-            .where_clause
-            .predicates
-            .last()
-            .map(|p| p.span())
-            .unwrap_or(end)
-            .shrink_to_hi()
-            .to(end),
+        generics.where_clause.tail_span_for_suggestion(),
         format!(
             "{} {}",
             if !generics.where_clause.predicates.is_empty() { "," } else { " where" },

--- a/src/librustc_trait_selection/traits/error_reporting/suggestions.rs
+++ b/src/librustc_trait_selection/traits/error_reporting/suggestions.rs
@@ -169,8 +169,17 @@ pub trait InferCtxtExt<'tcx> {
 }
 
 fn predicate_constraint(generics: &hir::Generics<'_>, pred: String) -> (Span, String) {
+    let end = generics.where_clause.span_for_predicates_or_empty_place().shrink_to_hi();
     (
-        generics.where_clause.span_for_predicates_or_empty_place().shrink_to_hi(),
+        // Account for `where T: Foo,` so we don't suggest two trailing commas.
+        generics
+            .where_clause
+            .predicates
+            .last()
+            .map(|p| p.span())
+            .unwrap_or(end)
+            .shrink_to_hi()
+            .to(end),
         format!(
             "{} {}",
             if !generics.where_clause.predicates.is_empty() { "," } else { " where" },

--- a/src/test/ui/bound-suggestions.rs
+++ b/src/test/ui/bound-suggestions.rs
@@ -19,7 +19,7 @@ fn test_one_bound<T: Sized>(t: T) {
 }
 
 #[allow(dead_code)]
-fn test_no_bounds_where<X, Y>(x: X, y: Y) where X: std::fmt::Debug {
+fn test_no_bounds_where<X, Y>(x: X, y: Y) where X: std::fmt::Debug, {
     println!("{:?} {:?}", x, y);
     //~^ ERROR doesn't implement
 }

